portal/portal.go

```diff
@@ -124,11 +124,50 @@ func ResolveLocalClientIP() (string, error) {
 	return conn.LocalAddr().(*net.UDPAddr).IP.String(), nil
 }
 
-// rsp struct for converting from raw response data to JSON
-type rsp struct {
-	ClientIP  string `json:"client_ip"`
-	Challenge string `json:"challenge"`
-	Error     string `json:"error"`
+// commonRsp struct for login session specific response
+type commonRsp struct {
+	// return code and various messages
+	// trash, but we have to add it
+	Status     string `json:"error"`
+	ErrorMsg   string `json:"error_msg"`
+	PloyMsg    string `json:"ploy_msg"`
+	SuccessMsg string `json:"suc_msg"`
+
+	// client_ip
+	ClientIP   string `json:"client_ip"`
+	// online_ip
+	OnlineIP   string `json:"online_ip"`
+	// challenge
+	Challenge  string `json:"challenge"`
+}
+
+// Error implements the error interface for commonRsp
+func (cr *commonRsp) Error() string {
+	// handle error msg and code based on priority
+	if cr.PloyMsg != "" {
+		return cr.PloyMsg
+	}
+	if cr.ErrorMsg != "" {
+		return cr.ErrorMsg
+	}
+	if cr.Status != "" && cr.Status != "ok" {
+		return cr.Status
+	}
+	// fallback
+	return "unknown portal response error"
+}
+
+// Err checks if the response indicates an error
+func (cr *commonRsp) Err() error {
```

Comment:
No need to be public here.
```suggestion
func (cr *commonRsp) error() error {
```